### PR TITLE
Introduce a new opentelemetry plugin for observability

### PR DESCRIPTION
**Pull request by Tantalor93**
### 1. Why is this pull request needed and what does it do?
This PR creates a new `opentelemetry` plugin, which should replace `trace` plugin in providing tracing (Observability) capabilities, it is built on top of the new standard [OpenTelemetry](https://opentelemetry.io/docs/), which replaces [OpenTracing](https://opentracing.io/) (OpenTracing standard is now being deprecated). The new plugin supports exporting traces into Zipkin collectors and OpenTelemetry collectors

### 2. Which issues (if any) are related?
#5530

### 3. Which documentation changes (if any) need to be made?
documentation of a new plugin

### 4. Does this introduce a backward incompatible change or deprecation?
will deprecate `trace` plugin

### CODE REVIEW

File: .gitignore
Comment (SuperQ) [inline .gitignore]: Please don't include your local editor ignores in the repo. Use [`git config --global core.excludesFile`](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/).

**Review by SuperQ — CHANGES_REQUESTED** [review]

**Comment by SuperQ** [comment]
I'm not sure I want opentelemetry in CoreDNS. I'd rather go with a lighter weight tracing library. OTel brings in a lot of excess baggage.

Perhaps this would be more appropriate as an external plugin.

File: plugin/opentelemetry/setup.go
Comment (mariuskimmina) [inline plugin/opentelemetry/setup.go]: Creating `var err error` once should be enough here I think.

File: plugin/opentelemetry/setup.go
Comment (mariuskimmina) [inline plugin/opentelemetry/setup.go]: I think we should use the official `OpenTelemetry` name with uppercase O and T.

**Review by mariuskimmina — COMMENTED** [review]
Overall this looks good to me, just some small comments. 

Should be able to try this out in the next few days and give further feedback

**Comment by Tantalor93** [comment]
@SuperQ I think I dont see much of an option to have this as external plugin as it needs to access (and modify) core functionality to be able to instrument plugin chain 🤔 (this https://github.com/Tantalor93/coredns/blob/01a9635a84890289ead663de1154efd126864406/plugin/plugin.go#L82)

**Comment by Tantalor93** [comment]
And I dont think that OpenTelemetry GO SDK is that heavyweight, based on go.mod changes, it basically adds few dependencies, where the biggest are zipkin and otel exporters, and as this new plugin will basically replace `trace` plugin, in future releases we could remove `trace` plugin altogether with all its OpenTracing and Datadog dependencies and make CoreDNS more lightweight. Or @SuperQ what makes you think that OpenTelemetry is too heavyweight?

**Comment by SuperQ** [comment]
While it doesn't look like much, the imports bring in about 15k LoC in dependencies based on a scan of the changes to `go mod vendor`. I would be interested in how this compares to simplifying our `trace` plugin to a more dedicated tracing library like [zipkin-go](https://github.com/openzipkin/zipkin-go).

**Comment by Tantalor93** [comment]
Simplifying trace plugin to depend only on zipkin-go might be a way, although OpenTelemetry as a observability standard is more extensible and would allow CoreDNS to work with a more type of collectors rather than just Zipkin (opentelemetry plugin in this state of PR is able to operate with both Zipkin and OpenTel collectors). Also adding OpenTelemetry into CoreDNS allows even third party plugins to further built and extend observability capabilities of their plugins on top of OpenTelemetry. Binding observability capabilities of CoreDNS only to one concrete technology (Zipkin) doesnt sound right to me.

**Comment by SuperQ** [comment]
> Also adding OpenTelemetry into CoreDNS allows even third party plugins to further built and extend observability capabilities of their plugins on top of OpenTelemetry.

This sounds a bit like a solution in search of a problem. What actual problems does this solve for CoreDNS?

For example, the existing Prometheusmetrics plugin allows all CoreDNS plugins to integrate into that observability framework.

Can other plugins not use the existing `trace` plugin?

> Binding observability capabilities of CoreDNS only to one concrete technology (Zipkin) doesnt sound right to me.

But this is just tracing. Is tracing really so complicated that we can't just produce a format that other systems can consume it?

Again, I feel like this is a solution in search of a problem. Why is this a separate plugin and not changes to the existing `trace` plugin?

**Comment by Tantalor93** [comment]
> This sounds a bit like a solution in search of a problem. What actual problems does this solve for CoreDNS?

This PR is trying to solve the problem that the OpenTracing project, upon which the `trace` plugin is built is now an archived project as [announced](https://github.com/opentracing/specification/issues/163) by maintainers of the project and also by CNCF ([announcement](https://www.cncf.io/blog/2022/01/31/cncf-archives-the-opentracing-project/)) and it is recommended to migrate to an OpenTelemetry

> But this is just tracing. Is tracing really so complicated that we can't just produce a format that other systems can consume it?

I think this is also one of the points of this PR, to add support for tracing in a new uniform tracing standard 

> Why is this a separate plugin and not changes to the existing trace plugin?

support for OpenTelemetry has been added as a separate plugin in order to maintain backward compatibility and to allow users of CoreDNS to migrate to an OpenTelemetry standard consciously

File: .gitignore
Comment (Tantalor93) [reply .gitignore]: good tip! removed

File: plugin/opentelemetry/setup.go
Comment (Tantalor93) [reply plugin/opentelemetry/setup.go]: you are right, fixed!

File: plugin/opentelemetry/setup.go
Comment (Tantalor93) [reply plugin/opentelemetry/setup.go]: yea, seems better, fixed!

**Comment by chrisohaver** [comment]
I agree that there is a clear need to move to OpenTelemetry, since OpenTracing has been deprecated. 

I'm not sure about new plugin vs updating the existing plugin. On the surface, it seems that updating the existing _trace_ plugin to use openTelemetry would be less disruptive to users unless it is truly backward incompatible. Can we leverage the [opentracing compatibility layer](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/opentracing.md)?

**Comment by mariuskimmina** [comment]
Got around to test it out a bit - I am not very familiar with OpenTelemetry, and yet it was easy to get this to work (with zipkin) - good job there!

**Comment by mariuskimmina** [comment]
One question: 
I used the following Corefile:
```
.:54 {
    log
    opentelemetry zipkin http://localhost:9411/api/v2/spans
    whoami
}
```

I am using `http` here without the `insecure` flag and it seems to work just fine - is that intended?
From your documentation it sounds like http should not be possible without the `insecure` flag

**Comment by Tantalor93** [comment]
@mariuskimmina yea that `insecure` flag is mainly to enable using OpenTelemetry exporter via HTTP, Zipkin is not that strict and works with HTTP even without an insecure option

@chrisohaver probably we could use that opentracing-opentelemetry bridge and migrate `trace` plugin in a backward compatible way (I would have to check this), although that will not enable us to get rid of OpenTracing dependency anytime soon. So my idea was to go around this and rather built a new plugin on top of a new SDK and have a plugin with a new cleaner API. While maintaining the old `trace` plugin for a few releases and then the drop `trace` plugin.

**Comment by chrisohaver** [comment]
Yes, the compatibility layer doesn't give us anything at this point, since the new implementation is already completed.

I think given that it's already implemented as a separate plugin at this point, it's unlikely we'll identify any reason compelling enough to warrant reworking this into the existing trace plugin.

**Comment by Tantalor93** [comment]
Are there any actions required from me or any questions that need to be answered in order to move forward with this PR? @chrisohaver @SuperQ or will we wait for more opinions from other maintainers? I just want to avoid this to go stale

**Comment by chrisohaver** [comment]
I’ll try to dedicate time to fully review this coming week.

**Comment by chrisohaver** [comment]
> I’ll try to dedicate time to fully review this coming week.

Actually I’ve got some older PRs in my queue, so I probably will not get to this one next week.

File: plugin/trace/README.md
Comment (chrisohaver) [inline plugin/trace/README.md]: I'm thinking we should deprecate _trace_ now, and remove it in the next minor release.

**Review by chrisohaver — APPROVED** [review]
LGTM

File: plugin/trace/README.md
Comment (Tantalor93) [reply plugin/trace/README.md]: ok, what should be done to deprecate the plugin @chrisohaver?

Comment (chrisohaver) [reply plugin/trace/README.md]: If others agree we should deprecate now, we just mention that the plugin is deprecated prominently in the plugin README. And in next release, we'd mention the plugin has been deprecated in the release notes.

Later in a future minor release, we'd remove the plugin.  If I recall, our policy is to wait at lease a couple of releases before removing a plugin/disabling a feature after deprecation.

Comment (Tantalor93) [reply plugin/trace/README.md]: I have added a deprecation notice into trace plugin README 👍

**Comment by Dentrax** [comment]
Great work!

Cc'ing @harshanarayana for awareness. 😇

**Comment by Tantalor93** [comment]
@chrisohaver I am wondering whether we should wait for more maintainers to approve/disapprove this PR or how to approach this PR? WDYT?

**Comment by chrisohaver** [comment]
> I'm not sure I want opentelemetry in CoreDNS. I'd rather go with a lighter weight tracing library. OTel brings in a lot of excess baggage.
> Perhaps this would be more appropriate as an external plugin.

@superq, have you changed your position on replacing the trace plugin with this opentelemetry plugin?

**Comment by chrisohaver** [comment]
Let's merge this.  @Tantalor93 can you update go mod/sum?

**Comment by SuperQ** [comment]
@chrisohaver No, not really. I'm not a fan of Open Telemetry as a project. I don't think it's going in the right direction. It's a huge kitchen sink of stuff.

**Comment by chrisohaver** [comment]
> @chrisohaver No, not really. I'm not a fan of Open Telemetry as a project. I don't think it's going in the right direction. It's a huge kitchen sink of stuff.

Thanks for responding. Can you suggest alternatives, path forward, what to do with existing trace plugin, and how to support tracing natively in CoreDNS in the future?

**Comment by SuperQ** [comment]
Like I posted above, [zipkin-go](https://github.com/openzipkin/zipkin-go) might be a good option. From what I've seen in tracing, zipkin is a reasonably supported format that can be accepted by most telemetry receivers. So even if someone wants to use OpenTelemetry, it will happily accept zipkin as the format.

**Comment by SuperQ** [comment]
And again, even if we do want to use OTel library, why is this a separate plugin and not an update to the existing tracing plugin?

**Comment by chrisohaver** [comment]
> And again, even if we do want to use OTel library, why is this a separate plugin and not an update to the existing tracing plugin?

If it was an update to the existing trace plugin, would that change your position on including it in coredns?

edit: FWIW, a reason for making it a separate plugin was given [above](https://github.com/coredns/coredns/pull/5777#issuecomment-1328233569), but I suspect this doesn't matter since you are opposed to including the opentel package in CoreDNS - I don't think it would matter which plugin includes it.

**Comment by chrisohaver** [comment]
> Like I posted above, [zipkin-go](https://github.com/openzipkin/zipkin-go) might be a good option. From what I've seen in tracing, zipkin is a reasonably supported format that can be accepted by most telemetry receivers. So even if someone wants to use OpenTelemetry, it will happily accept zipkin as the format.

Ok that sounds reasonable.

**Comment by thipokch** [comment]
Where can I find this external plugin?

**Comment by chrisohaver** [comment]
I don’t think there is one.

**Pull request closed without merging**
